components/cloud/gcp/src/kms.rs

```diff
@@ -0,0 +1,423 @@
+    #[serde(with = "serde_base64_bytes")]
+    ciphertext: Vec<u8>,
+    #[serde(deserialize_with = "deseralize_u32_from_str")]
+    ciphertext_crc32c: u32,
+}
+
+fn to_data_key(encrypt_resp: EncryptResp, raw_bytes: Vec<u8>) -> Result<DataKeyPair> {
+    Ok(DataKeyPair {
+        encrypted: EncryptedKey::new(encrypt_resp.ciphertext)?,
+        plaintext: PlainKey::new(raw_bytes, CryptographyType::AesGcm256)?,
+    })
+}
+
+#[derive(Clone, Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct DecryptRequest {
+    #[serde(with = "serde_base64_bytes")]
+    ciphertext: Vec<u8>,
+    ciphertext_crc32c: u32,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct DecryptResp {
+    #[serde(with = "serde_base64_bytes")]
+    plaintext: Vec<u8>,
+    #[serde(deserialize_with = "deseralize_u32_from_str")]
+    plaintext_crc32c: u32,
+}
+
+fn check_crc32(data: &[u8], expected: u32) -> StdResult<(), Crc32Error> {
+    let crc = crc32c::crc32c(data);
+    if crc != expected {
+        return Err(Crc32Error { expected, got: crc });
+    }
+    Ok(())
+}
+
+pub struct Crc32Error {
+    expected: u32,
+    got: u32,
+}
+
+impl fmt::Debug for Crc32Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+impl fmt::Display for Crc32Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "crc32 mismatch, expected: {}, got: {}",
+            self.expected, self.got
+        )
+    }
+}
+
+impl std::error::Error for Crc32Error {}
+
+impl From<Crc32Error> for CloudError {
+    fn from(e: Crc32Error) -> Self {
+        Self::KmsError(KmsError::Other(Box::new(e)))
+    }
+}
+
+mod serde_base64_bytes {
+    use serde::{Deserialize, Deserializer, Serializer};
+
+    // deserialize bytes from base64 encoded string.
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
+    where
+        D: Deserializer<'de>,
+        D::Error: serde::de::Error,
+    {
+        let v = String::deserialize(deserializer)?;
+        base64::decode(v)
+            .map_err(|e| serde::de::Error::custom(format!("base64 decode failed: {:?}", e,)))
+    }
+
+    // serialize bytes with base64 encoding.
+    pub fn serialize<S>(data: &Vec<u8>, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let str_data = base64::encode(data);
+        serializer.serialize_str(&str_data)
+    }
+}
+
+fn deseralize_u32_from_str<'de, D>(deserializer: D) -> StdResult<u32, D::Error>
+where
+    D: Deserializer<'de>,
+    D::Error: serde::de::Error,
+{
+    let v = String::deserialize(deserializer)?;
+    v.parse().map_err(|e| {
+        serde::de::Error::custom(format!("case crc32 string '{}' as u32 failed: {:?}", &v, e,))
+    })
+}
+
+#[derive(Clone, Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct GenRandomBytesReq {
+    length_bytes: usize,
+    // TODO: Change to enum
```

Comment:
Is it redundant ? Or still should keep tracking the remaining work？

Reply:
generateRandomBytes supports different [`protection_level`](https://cloud.google.com/kms/docs/reference/rest/v1/ProtectionLevel), but I see the implementation in azure always use `HSM`, so I think here we can just set  to `HSM` too. Thus, for coding convenience, just define it as string instead of enum since we always pass a const value here.

Reply:
Got it.